PWM Support

posborne opened the pull request:
This is my first crack at adding PWM support.  It is working on my LPC1768 board (that is the only board I have added code for at this point).  I haven't touched `platformtree!` with these initial changes as I expect that to be changing significantly in the short term.

Feedback very welcome.  Its probably important to get the Trait for PWM close to right so feedback there very appreciated.  Also, general comments on use of `ioregs!`, etc. is also very much appreciated as I will have to use it extensively for adding KLxx support if I move forward with that.

There are a couple of commits that we might consider backing out (linker script changes).  I have left those in there for now for discussion.

posborne commented:
Hmm, I may have to look over things a bit more closely.  At 100% duty cycle, it still looks like the signal is not quite being fully driven (only driving at 98%).

![image](https://cloud.githubusercontent.com/assets/41714/8018512/8918e0c8-0be5-11e5-86a3-2a6468dbbc41.png)

posborne commented:
Fixed the aforementioned issue, so should be ready for review.  PWM on the LPC17xx supports addition features (registering interrupts on match, etc.), but this implementation only support configuring a traditional single-PWM output (e.g. it lets you configure several single-edged pwm outputs).

mcoffin commented on src/hal/lpc17xx/layout.ld:
```diff
@@ -10,7 +10,7 @@ ENTRY(main)
 MEMORY
 {
-    rom(RX)   : ORIGIN = 0x00000000, LENGTH = 0x8000
+    rom(RX)   : ORIGIN = 0x00000000, LENGTH = 64K
```
Can we look up the docs on this before we merge it?

posborne replied:
Here's the [parametric search for the LPC17xx](http://www.nxp.com/parametrics/1403790745385/#/p=1,s=0,f=,c=,rpp=,fs=0,sc=c4c8d2,so=asc,es=), sorted by Flash size.  It looks like the smallest is the LPC1751FBD80 with 32KB of flash.  I can probably change it to be that more conservative value and still have plenty of flash space for any of the examples, even with debug.

mcoffin commented on src/hal/pwm.rs:
```diff
@@ -0,0 +1,59 @@
+    let pulsewidth_us = (adj_duty_cycle * self.get_period_us() as f32) as u32;
+    self.set_pulsewidth_us(pulsewidth_us)
+  }
+}
```
:+1:

posborne replied:
Looking back at this, I also intended to provide default implementations for `*_ms` variations of the functions as well (as is done, for instance, with timers).  Not really a significant issue as those could be added easily later.

mcoffin commented on examples/app_rgb_pwm_lpc17xx.rs:
```diff
@@ -0,0 +1,97 @@
+// option.  This file may not be copied, modified, or distributed
+// except according to those terms.
+#![feature(plugin, no_std, core, start)]
+#![crate_type="staticlib"]
```
We want to build examples as executables right now.

posborne replied:
Ok, I based mine off `app_uart`.  I will add a commit fixing all examples.

mcoffin commented on src/hal/pwm.rs:
```diff
@@ -0,0 +1,59 @@
+/// duty cycle in the case of a period change.
+pub trait PWMOutput {
+  /// set the period in microseconds
+  fn set_period_us(&mut self, period_us: u32);
```
Why `u32` specifically for period? This is generic code for all platforms, might want it to be `usize`.

bharrisau replied:
Limiting the maximum period to 256 us just because the native pointer size
is 8 bit seems a little limiting.
On 9 Jun 2015 9:48 am, "Matt Coffin" <notifications@github.com> wrote:

> In src/hal/pwm.rs
> <https://github.com/hackndev/zinc/pull/311#discussion_r31977205>:
>
> > +
> > +//! Common definitions for all PWM on all MCUs
> > +
> > +/// Trait for any Puluse Width Modulated output
> > +///
> > +/// This interface is inspired by the mbed `PWMOut` interface.
> > +///
> > +/// Note that on some MCUs, the period may be dictated by
> > +/// a timer shared by several differents PWMs.  It is not
> > +/// guaranteed that the period and pulsewidth specified
> > +/// will remain identical if modified elsewhere.  A proper
> > +/// implementaiton will still seek to maintain a similar
> > +/// duty cycle in the case of a period change.
> > +pub trait PWMOutput {
> > +  /// set the period in microseconds
> > +  fn set_period_us(&mut self, period_us: u32);
>
> Why u32 specifically for period? This is generic code for all platforms,
> might want it to be usize.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hackndev/zinc/pull/311/files#r31977205>.
>

mcoffin replied:
Oh. duh.

posborne replied:
Yeah, my interface is based on the mbed interface which I assumed was at least a reasonable starting point: https://developer.mbed.org/users/mbed_official/code/mbed/docs/cbbeb26dbd92//classmbed_1_1PwmOut.html

mcoffin replied:
:+1:

mcoffin commented on src/hal/lpc17xx/pwm.rs:
```diff
@@ -0,0 +1,294 @@
+
+#[allow(missing_docs)]
+#[derive(Clone, Copy)]
+pub struct PWM {
```
Even with `missing_docs` should still document the actual struct even if the values are implicitly documented

farcaller commented:
test this please

farcaller commented on examples/app_rgb_pwm_lpc17xx.rs:
```diff
@@ -0,0 +1,96 @@
+// Copyright 2015, Paul Osborne <osbpau@gmail.com>
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
```
Any chance this can be single-licensed under Apache so that we have similar terms for all the demo code?

posborne replied:
No problem, hadn't noticed the license difference for the demo code.

farcaller commented on src/hal/lpc17xx/pwm.rs:
```diff
@@ -0,0 +1,300 @@
+#[allow(missing_docs)]
+#[derive(Clone, Copy)]
+pub enum PWMChannel {
+  CHANNEL0 = 0,  // reserved for period
```
Why not name it in some other way given it's not useable?

farcaller commented on src/hal/lpc17xx/pwm.rs:
```diff
@@ -0,0 +1,300 @@
+    // setup match register to ticks per period on CH0
+    CHANNEL0.set_mr(pwm_us_to_ticks(self.period_us));
+
+    // TODO: recalculate other registers based on the new period?
```
Please own this todo.

farcaller commented on src/hal/lpc17xx/pwm.rs:
```diff
@@ -0,0 +1,300 @@
+  }
+
+  fn update_pulsewidth(&self) {
+    // TODO: mbed increases ticks by one if ticks for pulsewidth and period
```
Ditto.

farcaller commented on src/hal/lpc17xx/pwm.rs:
```diff
@@ -0,0 +1,300 @@
+  use volatile_cell::VolatileCell;
+  use core::ops::Drop;
+
+  ioregs!(PWM1 = {
```
Can you please refactor with placement (`@ addr`) and remove the extern and linker def?

farcaller replied:
Oh wait, #300 didn't land yet.

posborne commented:
Kinda forgot about this guy.  I'll try to update this branch based on suggestions so we can get this code merged.

posborne commented:
@farcaller @mcoffin I believe I have now (finally) made all the updates that were brought up previously and I have rebased on master.

farcaller commented:
Thanks for the work. The tests fail on native due to nightly issues I guess.

I'll try to take a deeper look into this a bit later this week, lots of code in here.

farcaller commented on src/hal/layout_common.ld:
```diff
@@ -2,6 +2,8 @@
 __aeabi_unwind_cpp_pr0 = abort;
 __aeabi_unwind_cpp_pr1 = abort;
 __aeabi_unwind_cpp_pr2 = abort;
+__exidx_start          = abort;
```
Those should be defined in sections, not as symbols. I guess placing them after `_edata` is reasonable, to some extent.

farcaller commented on src/hal/lpc17xx/layout.ld:
```diff
@@ -10,7 +10,7 @@ ENTRY(main)
 {
-    rom(RX)   : ORIGIN = 0x00000000, LENGTH = 0x8000
+    rom(RX)   : ORIGIN = 0x00000000, LENGTH = 32K
     ram(WAIL) : ORIGIN = 0x10000000, LENGTH = 0x2000
```
Can you update ram to human-readable too?

farcaller commented on src/hal/lpc17xx/pwm.rs:
```diff
@@ -0,0 +1,296 @@
+
+
+#[path="../../util/ioreg.rs"]
+#[macro_use] mod ioreg;
```
This pulls in the old ioreg, that is not used in here, right?

farcaller commented on src/hal/lpc17xx/pwm.rs:
```diff
@@ -0,0 +1,296 @@
+}
+
+impl PWMChannel {
+
```
Remove empty line

farcaller commented on src/hal/lpc17xx/pwm.rs:
```diff
@@ -0,0 +1,296 @@
+/// calculate the number of PWM clock ticks for the given number of microseconds
+fn pwm_us_to_ticks(us: u32) -> u32 {
+  let pwm_clock_frequency_hz: u32 = PWM1Clock.frequency();
+  let pwm_clock_frequency_mhz: u32 = pwm_clock_frequency_hz / 1_000_000;
```
I wonder if llvm does a proper divmod optimisation here.

farcaller commented on src/hal/lpc17xx/pwm.rs:
```diff
@@ -0,0 +1,296 @@
+}
+
+impl PWM {
+
```
Blank line

farcaller commented on src/hal/lpc17xx/pwm.rs:
```diff
@@ -0,0 +1,296 @@
+
+    let pwm = PWM {
+      channel: channel,
+      period_us: 20_000,  // 20ms is pretty common
```
Should be an argument to `new` I guess

farcaller commented on src/hal/lpc17xx/pwm.rs:
```diff
@@ -0,0 +1,296 @@
+
+  fn update_pulsewidth(&self) {
+    // TODO(posborne): mbed increases ticks by one if ticks for pulsewidth and period
+    //   are equal.  In testing, however, this caused problems.
```
second line indent is off. :disappointed: on the actual issue

farcaller commented on src/hal/lpc17xx/pwm.rs:
```diff
@@ -0,0 +1,296 @@
+      13 => pwmmr4r,  //= if set, reset on pwmmr4
+      14 => pwmmr4s,  //= if set, stop on pwmmr4
+
+      15=> pwmmr5i,  //= if set, interrupt on pwmmr5
```
missed a space

posborne commented:
This commit rebases on master, addresses comments from the PR, and includes some additional cleanup related to PWM.  Tested working on my lpc1768 again:

![blinking](http://i.giphy.com/f9MaJiVlRU5fq.gif)

farcaller commented:
:+1:
